// vscode-langservers-extracted/lib/markdown-language-server/protocol.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.onWatcherChange = exports.getReferencesToFileInWorkspace = exports.deleteFileWatcher = exports.createFileWatcher = exports.findFilesRequestTypes = exports.readDirectoryRequestType = exports.statFileRequestType = exports.readFileRequestType = exports.parseRequestType = void 0;

var vscode_languageserver_1 = require("vscode-languageserver"); // From server


exports.parseRequestType = new vscode_languageserver_1.RequestType('markdown/parse');
exports.readFileRequestType = new vscode_languageserver_1.RequestType('markdown/readFile');
exports.statFileRequestType = new vscode_languageserver_1.RequestType('markdown/statFile');
exports.readDirectoryRequestType = new vscode_languageserver_1.RequestType('markdown/readDirectory');
exports.findFilesRequestTypes = new vscode_languageserver_1.RequestType('markdown/findFiles');
exports.createFileWatcher = new vscode_languageserver_1.RequestType('markdown/createFileWatcher');
exports.deleteFileWatcher = new vscode_languageserver_1.RequestType('markdown/deleteFileWatcher'); // To server

exports.getReferencesToFileInWorkspace = new vscode_languageserver_1.RequestType('markdown/getReferencesToFileInWorkspace');
exports.onWatcherChange = new vscode_languageserver_1.RequestType('markdown/onWatcherChange');
