// vscode-langservers-extracted/lib/markdown-language-server/workspace.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

require("core-js/modules/es.array.slice.js");

require("core-js/modules/es.function.name.js");

require("core-js/modules/es.array.from.js");

require("core-js/modules/es.regexp.exec.js");

require("core-js/modules/es.symbol.js");

require("core-js/modules/es.symbol.description.js");

require("core-js/modules/es.symbol.iterator.js");

require("regenerator-runtime/runtime.js");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.map.js");

require("core-js/modules/es.object.to-string.js");

require("core-js/modules/es.string.iterator.js");

require("core-js/modules/web.dom-collections.iterator.js");

require("core-js/modules/es.array.map.js");

require("core-js/modules/es.promise.js");

require("core-js/modules/es.array.filter.js");

require("core-js/modules/es.array.concat.js");

require("core-js/modules/es.regexp.to-string.js");

require("core-js/modules/es.typed-array.uint8-array.js");

require("core-js/modules/es.typed-array.copy-within.js");

require("core-js/modules/es.typed-array.every.js");

require("core-js/modules/es.typed-array.fill.js");

require("core-js/modules/es.typed-array.filter.js");

require("core-js/modules/es.typed-array.find.js");

require("core-js/modules/es.typed-array.find-index.js");

require("core-js/modules/es.typed-array.for-each.js");

require("core-js/modules/es.typed-array.includes.js");

require("core-js/modules/es.typed-array.index-of.js");

require("core-js/modules/es.typed-array.iterator.js");

require("core-js/modules/es.typed-array.join.js");

require("core-js/modules/es.typed-array.last-index-of.js");

require("core-js/modules/es.typed-array.map.js");

require("core-js/modules/es.typed-array.reduce.js");

require("core-js/modules/es.typed-array.reduce-right.js");

require("core-js/modules/es.typed-array.reverse.js");

require("core-js/modules/es.typed-array.set.js");

require("core-js/modules/es.typed-array.slice.js");

require("core-js/modules/es.typed-array.some.js");

require("core-js/modules/es.typed-array.sort.js");

require("core-js/modules/es.typed-array.subarray.js");

require("core-js/modules/es.typed-array.to-locale-string.js");

require("core-js/modules/es.typed-array.to-string.js");

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VsCodeClientWorkspace = void 0;

var vscode_languageserver_1 = require("vscode-languageserver");

var vscode_languageserver_textdocument_1 = require("vscode-languageserver-textdocument");

var md = require("vscode-markdown-languageservice");

var vscode_uri_1 = require("vscode-uri");

var protocol = require("./protocol");

var arrays_1 = require("./util/arrays");

var file_1 = require("./util/file");

var limiter_1 = require("./util/limiter");

var resourceMap_1 = require("./util/resourceMap");

var schemes_1 = require("./util/schemes");

var VsCodeClientWorkspace = /*#__PURE__*/function () {
  function VsCodeClientWorkspace(connection, config, documents, notebooks, logger) {
    var _this = this;

    _classCallCheck(this, VsCodeClientWorkspace);

    this.connection = connection;
    this.config = config;
    this.documents = documents;
    this.notebooks = notebooks;
    this.logger = logger;
    this._onDidCreateMarkdownDocument = new vscode_languageserver_1.Emitter();
    this.onDidCreateMarkdownDocument = this._onDidCreateMarkdownDocument.event;
    this._onDidChangeMarkdownDocument = new vscode_languageserver_1.Emitter();
    this.onDidChangeMarkdownDocument = this._onDidChangeMarkdownDocument.event;
    this._onDidDeleteMarkdownDocument = new vscode_languageserver_1.Emitter();
    this.onDidDeleteMarkdownDocument = this._onDidDeleteMarkdownDocument.event;
    this._documentCache = new resourceMap_1.ResourceMap();
    this._utf8Decoder = new TextDecoder('utf-8');
    this._watcherPool = 0;
    this._watchers = new Map();
    this._workspaceFolders = [];
    documents.onDidOpen(function (e) {
      _this._documentCache.delete(vscode_uri_1.URI.parse(e.document.uri));

      if (_this.isRelevantMarkdownDocument(e.document)) {
        _this._onDidCreateMarkdownDocument.fire(e.document);
      }
    });
    documents.onDidChangeContent(function (e) {
      if (_this.isRelevantMarkdownDocument(e.document)) {
        _this._onDidChangeMarkdownDocument.fire(e.document);
      }
    });
    documents.onDidClose(function (e) {
      _this._documentCache.delete(vscode_uri_1.URI.parse(e.document.uri));
    });
    connection.onDidChangeWatchedFiles( /*#__PURE__*/function () {
      var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref) {
        var changes, _iterator, _step, change, resource, document, _document;

        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                changes = _ref.changes;
                _iterator = _createForOfIteratorHelper(changes);
                _context.prev = 2;

                _iterator.s();

              case 4:
                if ((_step = _iterator.n()).done) {
                  _context.next = 26;
                  break;
                }

                change = _step.value;
                resource = vscode_uri_1.URI.parse(change.uri);
                _context.t0 = change.type;
                _context.next = _context.t0 === vscode_languageserver_1.FileChangeType.Changed ? 10 : _context.t0 === vscode_languageserver_1.FileChangeType.Created ? 16 : _context.t0 === vscode_languageserver_1.FileChangeType.Deleted ? 21 : 24;
                break;

              case 10:
                _this._documentCache.delete(resource);

                _context.next = 13;
                return _this.openMarkdownDocument(resource);

              case 13:
                document = _context.sent;

                if (document) {
                  _this._onDidChangeMarkdownDocument.fire(document);
                }

                return _context.abrupt("break", 24);

              case 16:
                _context.next = 18;
                return _this.openMarkdownDocument(resource);

              case 18:
                _document = _context.sent;

                if (_document) {
                  _this._onDidCreateMarkdownDocument.fire(_document);
                }

                return _context.abrupt("break", 24);

              case 21:
                _this._documentCache.delete(resource);

                _this._onDidDeleteMarkdownDocument.fire(resource);

                return _context.abrupt("break", 24);

              case 24:
                _context.next = 4;
                break;

              case 26:
                _context.next = 31;
                break;

              case 28:
                _context.prev = 28;
                _context.t1 = _context["catch"](2);

                _iterator.e(_context.t1);

              case 31:
                _context.prev = 31;

                _iterator.f();

                return _context.finish(31);

              case 34:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, null, [[2, 28, 31, 34]]);
      }));

      return function (_x) {
        return _ref2.apply(this, arguments);
      };
    }());
    connection.onRequest(protocol.onWatcherChange, function (params) {
      var watcher = _this._watchers.get(params.id);

      if (!watcher) {
        return;
      }

      switch (params.kind) {
        case 'create':
          watcher.onDidCreate.fire(vscode_uri_1.URI.parse(params.uri));
          return;

        case 'change':
          watcher.onDidChange.fire(vscode_uri_1.URI.parse(params.uri));
          return;

        case 'delete':
          watcher.onDidDelete.fire(vscode_uri_1.URI.parse(params.uri));
          return;
      }
    });
  }

  _createClass(VsCodeClientWorkspace, [{
    key: "listen",
    value: function listen() {
      var _this2 = this;

      this.connection.workspace.onDidChangeWorkspaceFolders( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
        var _yield$_this2$connect;

        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return _this2.connection.workspace.getWorkspaceFolders();

              case 2:
                _context2.t1 = _yield$_this2$connect = _context2.sent;
                _context2.t0 = _context2.t1 !== null;

                if (!_context2.t0) {
                  _context2.next = 6;
                  break;
                }

                _context2.t0 = _yield$_this2$connect !== void 0;

              case 6:
                if (!_context2.t0) {
                  _context2.next = 10;
                  break;
                }

                _context2.t2 = _yield$_this2$connect;
                _context2.next = 11;
                break;

              case 10:
                _context2.t2 = [];

              case 11:
                _this2.workspaceFolders = _context2.t2.map(function (x) {
                  return vscode_uri_1.URI.parse(x.uri);
                });

              case 12:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      })));
    }
  }, {
    key: "workspaceFolders",
    get: function get() {
      return this._workspaceFolders;
    },
    set: function set(value) {
      this._workspaceFolders = value;
    }
  }, {
    key: "getAllMarkdownDocuments",
    value: function () {
      var _getAllMarkdownDocuments = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4() {
        var _this3 = this;

        var maxConcurrent, foundFiles, limiter, resources, onDiskResults, openTextDocumentResults;
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                maxConcurrent = 20;
                foundFiles = new resourceMap_1.ResourceMap();
                limiter = new limiter_1.Limiter(maxConcurrent); // Add files on disk

                _context4.next = 5;
                return this.connection.sendRequest(protocol.findFilesRequestTypes, {});

              case 5:
                resources = _context4.sent;
                _context4.next = 8;
                return Promise.all(resources.map(function (strResource) {
                  return limiter.queue( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3() {
                    var resource, doc;
                    return regeneratorRuntime.wrap(function _callee3$(_context3) {
                      while (1) {
                        switch (_context3.prev = _context3.next) {
                          case 0:
                            resource = vscode_uri_1.URI.parse(strResource);
                            _context3.next = 3;
                            return _this3.openMarkdownDocument(resource);

                          case 3:
                            doc = _context3.sent;

                            if (doc) {
                              foundFiles.set(resource);
                            }

                            return _context3.abrupt("return", doc);

                          case 6:
                          case "end":
                            return _context3.stop();
                        }
                      }
                    }, _callee3);
                  })));
                }));

              case 8:
                onDiskResults = _context4.sent;
                _context4.next = 11;
                return Promise.all(this.documents.all().filter(function (doc) {
                  return !foundFiles.has(vscode_uri_1.URI.parse(doc.uri)) && _this3.isRelevantMarkdownDocument(doc);
                }));

              case 11:
                openTextDocumentResults = _context4.sent;
                return _context4.abrupt("return", (0, arrays_1.coalesce)([].concat(_toConsumableArray(onDiskResults), _toConsumableArray(openTextDocumentResults))));

              case 13:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getAllMarkdownDocuments() {
        return _getAllMarkdownDocuments.apply(this, arguments);
      }

      return getAllMarkdownDocuments;
    }()
  }, {
    key: "hasMarkdownDocument",
    value: function hasMarkdownDocument(resource) {
      return !!this.documents.get(resource.toString());
    }
  }, {
    key: "openMarkdownDocument",
    value: function () {
      var _openMarkdownDocument = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(resource) {
        var existing, matchingDocument, response, bytes, text, doc;
        return regeneratorRuntime.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                existing = this._documentCache.get(resource);

                if (!existing) {
                  _context5.next = 3;
                  break;
                }

                return _context5.abrupt("return", existing);

              case 3:
                matchingDocument = this.documents.get(resource.toString());

                if (!matchingDocument) {
                  _context5.next = 7;
                  break;
                }

                this._documentCache.set(resource, matchingDocument);

                return _context5.abrupt("return", matchingDocument);

              case 7:
                if ((0, file_1.looksLikeMarkdownPath)(this.config, resource)) {
                  _context5.next = 9;
                  break;
                }

                return _context5.abrupt("return", undefined);

              case 9:
                _context5.prev = 9;
                _context5.next = 12;
                return this.connection.sendRequest(protocol.readFileRequestType, {
                  uri: resource.toString()
                });

              case 12:
                response = _context5.sent;
                // TODO: LSP doesn't seem to handle Array buffers well
                bytes = new Uint8Array(response); // We assume that markdown is in UTF-8

                text = this._utf8Decoder.decode(bytes);
                doc = vscode_languageserver_textdocument_1.TextDocument.create(resource.toString(), 'markdown', 0, text);

                this._documentCache.set(resource, doc);

                return _context5.abrupt("return", doc);

              case 20:
                _context5.prev = 20;
                _context5.t0 = _context5["catch"](9);
                return _context5.abrupt("return", undefined);

              case 23:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this, [[9, 20]]);
      }));

      function openMarkdownDocument(_x2) {
        return _openMarkdownDocument.apply(this, arguments);
      }

      return openMarkdownDocument;
    }()
  }, {
    key: "stat",
    value: function () {
      var _stat = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6(resource) {
        return regeneratorRuntime.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                this.logger.log(md.LogLevel.Trace, 'VsCodeClientWorkspace: stat', "".concat(resource));

                if (!(this._documentCache.has(resource) || this.documents.get(resource.toString()))) {
                  _context6.next = 3;
                  break;
                }

                return _context6.abrupt("return", {
                  isDirectory: false
                });

              case 3:
                return _context6.abrupt("return", this.connection.sendRequest(protocol.statFileRequestType, {
                  uri: resource.toString()
                }));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function stat(_x3) {
        return _stat.apply(this, arguments);
      }

      return stat;
    }()
  }, {
    key: "readDirectory",
    value: function () {
      var _readDirectory = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee7(resource) {
        return regeneratorRuntime.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                this.logger.log(md.LogLevel.Trace, 'VsCodeClientWorkspace: readDir', "".concat(resource));
                return _context7.abrupt("return", this.connection.sendRequest(protocol.readDirectoryRequestType, {
                  uri: resource.toString()
                }));

              case 2:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function readDirectory(_x4) {
        return _readDirectory.apply(this, arguments);
      }

      return readDirectory;
    }()
  }, {
    key: "getContainingDocument",
    value: function getContainingDocument(resource) {
      if (resource.scheme === schemes_1.Schemes.notebookCell) {
        var nb = this.notebooks.findNotebookDocumentForCell(resource.toString());

        if (nb) {
          return {
            uri: vscode_uri_1.URI.parse(nb.uri),
            children: nb.cells.map(function (cell) {
              return {
                uri: vscode_uri_1.URI.parse(cell.document)
              };
            })
          };
        }
      }

      return undefined;
    }
  }, {
    key: "watchFile",
    value: function watchFile(resource, options) {
      var _this4 = this;

      var entry = {
        resource: resource,
        options: options,
        onDidCreate: new vscode_languageserver_1.Emitter(),
        onDidChange: new vscode_languageserver_1.Emitter(),
        onDidDelete: new vscode_languageserver_1.Emitter()
      };
      var id = this._watcherPool++;

      this._watchers.set(id, entry);

      this.connection.sendRequest(protocol.createFileWatcher, {
        id: id,
        uri: resource.toString(),
        options: options
      });
      return {
        onDidCreate: entry.onDidCreate.event,
        onDidChange: entry.onDidChange.event,
        onDidDelete: entry.onDidDelete.event,
        dispose: function dispose() {
          _this4.connection.sendRequest(protocol.deleteFileWatcher, {
            id: id
          });

          _this4._watchers.delete(id);
        }
      };
    }
  }, {
    key: "isRelevantMarkdownDocument",
    value: function isRelevantMarkdownDocument(doc) {
      return (0, file_1.isMarkdownFile)(doc) && vscode_uri_1.URI.parse(doc.uri).scheme !== 'vscode-bulkeditpreview';
    }
  }]);

  return VsCodeClientWorkspace;
}();

exports.VsCodeClientWorkspace = VsCodeClientWorkspace;
