// vscode-markdown-languageservice/out/index.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLanguageService = exports.githubSlugifier = exports.LogLevel = exports.DiagnosticLevel = exports.DiagnosticCode = void 0;
const config_1 = require("./config");
const definitions_1 = require("./languageFeatures/definitions");
const diagnostics_1 = require("./languageFeatures/diagnostics");
const documentLinks_1 = require("./languageFeatures/documentLinks");
const documentSymbols_1 = require("./languageFeatures/documentSymbols");
const fileRename_1 = require("./languageFeatures/fileRename");
const folding_1 = require("./languageFeatures/folding");
const pathCompletions_1 = require("./languageFeatures/pathCompletions");
const references_1 = require("./languageFeatures/references");
const rename_1 = require("./languageFeatures/rename");
const smartSelect_1 = require("./languageFeatures/smartSelect");
const workspaceSymbols_1 = require("./languageFeatures/workspaceSymbols");
const tableOfContents_1 = require("./tableOfContents");
const workspace_1 = require("./workspace");
var diagnostics_2 = require("./languageFeatures/diagnostics");
Object.defineProperty(exports, "DiagnosticCode", { enumerable: true, get: function () { return diagnostics_2.DiagnosticCode; } });
Object.defineProperty(exports, "DiagnosticLevel", { enumerable: true, get: function () { return diagnostics_2.DiagnosticLevel; } });
var logging_1 = require("./logging");
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return logging_1.LogLevel; } });
var slugify_1 = require("./slugify");
Object.defineProperty(exports, "githubSlugifier", { enumerable: true, get: function () { return slugify_1.githubSlugifier; } });
/**
 * Create a new instance of the language service.
 */
function createLanguageService(init) {
    const config = (0, config_1.getLsConfiguration)(init);
    const logger = init.logger;
    const tocProvider = new tableOfContents_1.MdTableOfContentsProvider(init.parser, init.workspace, logger);
    const docSymbolProvider = new documentSymbols_1.MdDocumentSymbolProvider(tocProvider, logger);
    const smartSelectProvider = new smartSelect_1.MdSelectionRangeProvider(init.parser, tocProvider, logger);
    const foldingProvider = new folding_1.MdFoldingProvider(init.parser, tocProvider, logger);
    const workspaceSymbolProvider = new workspaceSymbols_1.MdWorkspaceSymbolProvider(init.workspace, docSymbolProvider);
    const linkProvider = new documentLinks_1.MdLinkProvider(init.parser, init.workspace, tocProvider, logger);
    const pathCompletionProvider = new pathCompletions_1.MdPathCompletionProvider(config, init.workspace, init.parser, linkProvider);
    const linkCache = (0, documentLinks_1.createWorkspaceLinkCache)(init.parser, init.workspace);
    const referencesProvider = new references_1.MdReferencesProvider(config, init.parser, init.workspace, tocProvider, linkCache, logger);
    const definitionsProvider = new definitions_1.MdDefinitionProvider(config, init.workspace, tocProvider, linkCache);
    const renameProvider = new rename_1.MdRenameProvider(config, init.workspace, referencesProvider, init.parser.slugifier);
    const fileRenameProvider = new fileRename_1.MdFileRenameProvider(config, init.workspace, linkCache, referencesProvider);
    const diagnosticsComputer = new diagnostics_1.DiagnosticComputer(config, init.workspace, linkProvider, tocProvider);
    return Object.freeze({
        dispose: () => {
            linkCache.dispose();
            tocProvider.dispose();
            workspaceSymbolProvider.dispose();
            linkProvider.dispose();
            referencesProvider.dispose();
        },
        getDocumentLinks: linkProvider.provideDocumentLinks.bind(linkProvider),
        resolveDocumentLink: linkProvider.resolveDocumentLink.bind(linkProvider),
        getDocumentSymbols: docSymbolProvider.provideDocumentSymbols.bind(docSymbolProvider),
        getFoldingRanges: foldingProvider.provideFoldingRanges.bind(foldingProvider),
        getSelectionRanges: smartSelectProvider.provideSelectionRanges.bind(smartSelectProvider),
        getWorkspaceSymbols: workspaceSymbolProvider.provideWorkspaceSymbols.bind(workspaceSymbolProvider),
        getCompletionItems: pathCompletionProvider.provideCompletionItems.bind(pathCompletionProvider),
        getReferences: referencesProvider.provideReferences.bind(referencesProvider),
        getFileReferences: async (resource, token) => {
            return (await referencesProvider.getReferencesToFileInWorkspace(resource, token)).map(x => x.location);
        },
        getDefinition: definitionsProvider.provideDefinition.bind(definitionsProvider),
        prepareRename: renameProvider.prepareRename.bind(renameProvider),
        getRenameEdit: renameProvider.provideRenameEdits.bind(renameProvider),
        getRenameFilesInWorkspaceEdit: fileRenameProvider.getRenameFilesInWorkspaceEdit.bind(fileRenameProvider),
        computeDiagnostics: async (doc, options, token) => {
            return (await diagnosticsComputer.compute(doc, options, token))?.diagnostics;
        },
        createPullDiagnosticsManager: () => {
            if (!(0, workspace_1.isWorkspaceWithFileWatching)(init.workspace)) {
                throw new Error(`Workspace does not support file watching. Diagnostics manager not supported`);
            }
            return new diagnostics_1.DiagnosticsManager(config, init.workspace, linkProvider, tocProvider);
        }
    });
}
exports.createLanguageService = createLanguageService;
